rig/bitfield.py

```diff
@@ -361,19 +364,60 @@ def assign_fields(self):
         Users should typically call this method after all field values have
         been assigned, otherwise fields may be fixed at an inadequate size.
         """
-        # We must fix fields at every level of the heirarchy sepeartely
+        # We must fix fields at every level of the hierarchy separately
         # (otherwise fields of children won't be allowed to overlap). Here we
-        # do a breadth-first iteration over the heirarchy, fixing the fields at
-        # each level.
-        unsearched_heirarchy = [BitField(self.length, self.fields)]
-        while unsearched_heirarchy:
-            ks = unsearched_heirarchy.pop(0)
-            ks._assign_enabled_fields()
-            # Look for potential children in the herarchy
+        # do a breadth-first iteration over the hierarchy to fix fields with
+        # given starting positions; then we do depth-first to fix other fields.
+        #
+        # The breadth-first ensures that children's fixed position fields must
+        # fit around the fixed position fields of their parents.  The depth
+        # first search for variable position fields ensures that parents don't
+        # allocate fields in positions which would collide with fixed and
+        # variable position fields their children have already allocated.
+
+        class Node(namedtuple("Node", "bitfield children")):
+            """Node used in depth-first search of bitfields."""
+            def recurse_assign_unfixed_fields(self, assigned_bits=0,
+                                              excluded_fields=set()):
+                # Call this for the children first
+                excludes = set(i for (i, f) in self.bitfield._enabled_fields())
+                new_assigned_bits = assigned_bits
+                for child in self.children:
+                    # Children can assign bits independently
+                    new_assigned_bits |= child.recurse_assign_unfixed_fields(
+                        assigned_bits, excludes)
+
+                # Now assign locally
+                assigned_bits =\
+                    self.bitfield._assign_enabled_fields_without_fixed_start(
+                        new_assigned_bits, excluded_fields
+                    )
+
+                # Check that everything HAS been assigned
```

Comment:
Is there some user-error that can cause this to happen or is this just out of lack of certainty that the algorithm works? (If the former, jolly good, if the latter...)